great_expectations/core/util.py

```diff
@@ -782,148 +771,150 @@ def sniff_s3_compression(s3_url: S3Url) -> Union[str, None]:
-        # noinspection PyProtectedMember,PyUnresolvedReferences
-        if spark_session.sparkContext._jsc.sc().isStopped():
-            raise ValueError("SparkContext stopped unexpectedly.")
-
-    except AttributeError:
-        logger.error(
-            "Unable to load spark context; install optional spark dependency for support."
-        )
-        spark_session = None
-
-    return spark_session
+        app_name: str = spark_session.sparkContext.appName  # type: ignore[assignment]  # handled by try/except
+    except pyspark.PySparkNotImplementedError:
+        # if the session is a Spark/Databricks Connect session,
+        # this error is raised and the config is not updatable
+        return False
 
+    updatable = (app_name == DATABRICKS_SHELL_APP_NAME) and (
+        os.environ.get("DATABRICKS_RUNTIME_VERSION") is not None  # noqa: TID251
+    )
+    return updatable
 
-def spark_restart_required(
-    current_spark_config: List[Tuple[str, Any]], desired_spark_config: dict
-) -> bool:
-    """Determines whether or not Spark session should be restarted, based on supplied current and desired configuration.
 
-    Either new "App" name or configuration change necessitates Spark session restart.
+def _get_session_with_spark_config(
+    spark_session: pyspark.SparkSession,
+    spark_config: dict,
+    allow_restart: bool,
+) -> pyspark.SparkSession:
+    """Gets a SparkSession with spark_config by updating or re-starting an existing SparkSession.
 
     Args:
-        current_spark_config: List of tuples containing Spark configuration string-valued key/property pairs.
-        desired_spark_config: List of tuples containing Spark configuration string-valued key/property pairs.
+        spark_session: An existing PySpark SparkSession.
+        spark_config: A dictionary of SparkSession.Builder.config objects.
+        allow_restart: Whether to restart the existing SparkSession.
 
-    Returns: Boolean flag indicating (if True) that Spark session restart is required.
+    Returns:
+        SparkSession
     """
-
-    # we can't change spark context config values within databricks runtimes
-    if in_databricks():
-        return False
-
-    current_spark_config_dict: dict = {k: v for (k, v) in current_spark_config}
-    if desired_spark_config.get("spark.app.name") != current_spark_config_dict.get(
-        "spark.app.name"
-    ):
-        return True
-
-    if not {(k, v) for k, v in desired_spark_config.items()}.issubset(
-        current_spark_config
-    ):
-        return True
-
-    return False
+    if allow_restart:
+        try:
+            retrieved_spark_config: pyspark.SparkConf = (
+                spark_session.sparkContext.getConf()
+            )
+            for key, value in spark_config.items():
+                if retrieved_spark_config.get(key) != value:
+                    spark_session.stop()
+                    break
+        except pyspark.PySparkNotImplementedError:
```

Comment:
What about the security error raised by getConf if we're in a shared compute cluster? Are we sure that we've always caught that with allow_restart? Should we catch the security condition?

Reply:
This is outdated now. We are handling the security error in the revisions.